### PR TITLE
cxxrtl: improve robustness of `cxxrtl::time`

### PR DESCRIPTION
**Pull request by whitequark**
Avoid overflow during conversion for any representable raw value.

### CODE REVIEW

File: backends/cxxrtl/runtime/cxxrtl/cxxrtl_time.h [hunk 1]
```diff
@@ -68,12 +70,12 @@ class time {
 		return raw.is_neg();
 	}
 
-	// Extracts the absolute number of whole seconds. Negative if the value is negative.
-	int32_t secs() const {
-		return raw.sdivmod(resolution).first.trunc<32>().get<uint32_t>();
+	// Extracts the number of whole seconds. Negative if the value is negative.
+	int64_t secs() const {
+		return raw.sdivmod(resolution).first.trunc<64>().get<uint64_t>();
```
Comment (povik) [inline backends/cxxrtl/runtime/cxxrtl/cxxrtl_time.h:75]: `.get<uint64_t>` vs `int64_t secs()` intentional?

**Review by povik — APPROVED** [review]
I have a better feeling about it now, looks good to me.

File: backends/cxxrtl/runtime/cxxrtl/cxxrtl_time.h (hunk 1 above)
Comment (whitequark) [reply backends/cxxrtl/runtime/cxxrtl/cxxrtl_time.h:75]: `.get<int64_t>()` isn't implemented.

Comment (whitequark) [reply backends/cxxrtl/runtime/cxxrtl/cxxrtl_time.h:75]: Oh, wait... I need a `bit_cast` here don't I?

**Comment by povik** [comment]
Failure from CI:
```
 + gcc-11 -std=c++11 -o yosys-always_full -I../../backends/cxxrtl/runtime always_full_tb.cc -lstdc++
In file included from yosys-always_full.cc:1,
                 from always_full_tb.cc:1:
../../backends/cxxrtl/runtime/cxxrtl/cxxrtl.h: In member function ‘IntegerT cxxrtl::value<Bits>::get() const’:
../../backends/cxxrtl/runtime/cxxrtl/cxxrtl.h:165:49: error: ‘make_unsigned_t’ is not a member of ‘std’; did you mean ‘make_unsigned’?
  165 |                 auto unsigned_result = get<std::make_unsigned_t<IntegerT>>();
      |                                                 ^~~~~~~~~~~~~~~
      |                                                 make_unsigned
../../backends/cxxrtl/runtime/cxxrtl/cxxrtl.h:165:49: error: ‘make_unsigned_t’ is not a member of ‘std’; did you mean ‘make_unsigned’?
  165 |                 auto unsigned_result = get<std::make_unsigned_t<IntegerT>>();
      |                                                 ^~~~~~~~~~~~~~~
      |                                                 make_unsigned
../../backends/cxxrtl/runtime/cxxrtl/cxxrtl.h:165:40: error: parse error in template argument list
  165 |                 auto unsigned_result = get<std::make_unsigned_t<IntegerT>>();
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../backends/cxxrtl/runtime/cxxrtl/cxxrtl.h:165:76: error: expected primary-expression before ‘)’ token
  165 |                 auto unsigned_result = get<std::make_unsigned_t<IntegerT>>();
      |                                                                            ^
../../backends/cxxrtl/runtime/cxxrtl/cxxrtl.h: In member function ‘void cxxrtl::value<Bits>::set(IntegerT)’:
../../backends/cxxrtl/runtime/cxxrtl/cxxrtl.h:185:22: error: ‘make_unsigned_t’ is not a member of ‘std’; did you mean ‘make_unsigned’?
  185 |                 std::make_unsigned_t<IntegerT> unsigned_value;
      |                      ^~~~~~~~~~~~~~~
      |                      make_unsigned
../../backends/cxxrtl/runtime/cxxrtl/cxxrtl.h:185:46: error: expected primary-expression before ‘>’ token
  185 |                 std::make_unsigned_t<IntegerT> unsigned_value;
      |                                              ^
../../backends/cxxrtl/runtime/cxxrtl/cxxrtl.h:185:48: error: ‘unsigned_value’ was not declared in this scope
  185 |                 std::make_unsigned_t<IntegerT> unsigned_value;
      |                                                ^~~~~~~~~~~~~~
```
This is with c++11.

**Comment by povik** [comment]
Maybe std::make_unsigned<T>::type is an option?

**Comment by whitequark** [comment]
Yeah, it should be `typename std::make_unsigned<>::type` I think

**Comment by whitequark** [comment]
Done

**Review by povik — APPROVED** [review]
I was surprised to see those `memcpy`s there but spending a moment on http://cppreference.com/ I think I understand. (Though that's still technically implementation-defined, no?)

**Comment by whitequark** [comment]
> Though that's still technically implementation-defined, no?

Yes, but with C++20 making 2's complement mandatory, I don't really care.

**Pull request merged**